src/python-samplerate.cpp

```diff
@@ -0,0 +1,570 @@
+    throw ResamplingException(errnum);
+  } else if (errnum != 0) {  // the zero case is excluded as it is not an error
+    // this will throw a segmentation fault if we call src_strerror here
+    // also, these should never happen
+    throw std::runtime_error("libsamplerate raised an unknown error code");
+  }
+}
+
+class Resampler {
+ private:
+  SRC_STATE *_state = nullptr;
+  int _err_num = 0;
+
+ public:
+  int _converter_type = 0;
+  int _channels = 0;
+
+ public:
+  Resampler(const py::object &converter_type, int channels)
+      : _converter_type(get_converter_type(converter_type)),
+        _channels(channels) {
+    _state = src_new(_converter_type, _channels, &_err_num);
+  }
+
+  // copy constructor
+  Resampler(const Resampler &r)
+      : _converter_type(r._converter_type), _channels(r._channels) {
+    _state = src_clone(r._state, &_err_num);
+    error_handler(_err_num);
+  }
+
+  ~Resampler() { src_delete(_state); }
+
+  py::array_t<float, py::array::c_style> process(
+      py::array_t<float, py::array::c_style | py::array::forcecast> input,
+      double sr_ratio, bool end_of_input) {
+    // accessors for the arrays
+    py::buffer_info inbuf = input.request();
+
+    // set the number of channels
+    int channels = 1;
+    if (inbuf.ndim == 2)
+      channels = inbuf.shape[1];
+    else if (inbuf.ndim > 2)
+      throw std::domain_error("Input array should have at most 2 dimensions");
+
+    if (channels != _channels || channels == 0)
+      throw std::domain_error("Invalid number of channels in input data.");
+
+    const size_t new_size = size_t(std::ceil(inbuf.shape[0] * sr_ratio));
+
+    // allocate output array
+    std::vector<size_t> out_shape{new_size};
+    if (inbuf.ndim == 2) out_shape.push_back(size_t(channels));
+    auto output = py::array_t<float, py::array::c_style>(out_shape);
+    py::buffer_info outbuf = output.request();
+
+    // libsamplerate struct
+    SRC_DATA src_data = {
+        static_cast<float *>(inbuf.ptr),   // data_in
+        static_cast<float *>(outbuf.ptr),  // data_out
+        inbuf.shape[0],                    // input_frames
+        long(new_size),                    // output_frames
+        0,             // input_frames_used, filled by libsamplerate
+        0,             // output_frames_gen, filled by libsamplerate
+        end_of_input,  // end_of_input, not used by src_simple ?
+        sr_ratio       // src_ratio, sampling rate conversion ratio
+    };
+
+    _err_num = src_process(_state, &src_data);
+    error_handler(_err_num);
+
+    // create a shorter view of the array
+    if ((size_t)src_data.output_frames_gen < new_size) {
+      out_shape[0] = src_data.output_frames_gen;
+      output = py::array_t<float, py::array::c_style>(
+          out_shape, static_cast<float *>(outbuf.ptr));
+    }
+
+    return output;
+  }
+
+  void set_ratio(double new_ratio) {
+    _err_num = src_set_ratio(_state, new_ratio);
+    error_handler(_err_num);
```

Comment:
Why actually keep `_err_num` around? It could be a temporary,
```suggestion
    handle_errors(src_set_ratio(_state, new_ratio));
```